File: android/app/src/main/java/com/openburnbar/data/policy/MobileHermesConversationPolicy.kt
package com.openburnbar.data.policy

enum class MobileHermesStreamTerminal(val wire: String) {
    STREAMING("streaming"),
    STOPPED("stopped"),
    CANCELLED("cancelled"),
    ERROR("error"),
    COMPLETED("completed"),
}

enum class MobileHermesAttachmentDisposition(val wire: String) {
    ACCEPTED("accepted"),
    REJECTED("rejected"),
}

enum class MobileHermesConversationDeepLink(val wire: String) {
    LOADED("loaded"),
    MISSING("missing"),
    INVALID("invalid"),
}

/** Hermes/Pi stream, attachment, and assistant-link decisions. Source: iOS cancelStreaming. */
object MobileHermesConversationPolicy {
    fun terminal(event: String): MobileHermesStreamTerminal = when (event) {
        "stop" -> MobileHermesStreamTerminal.STOPPED
        "cancel" -> MobileHermesStreamTerminal.CANCELLED
        "error" -> MobileHermesStreamTerminal.ERROR
        "complete" -> MobileHermesStreamTerminal.COMPLETED
        else -> MobileHermesStreamTerminal.ERROR
    }

    fun keepsPartial(terminal: MobileHermesStreamTerminal): Boolean = when (terminal) {
        MobileHermesStreamTerminal.STOPPED,
        MobileHermesStreamTerminal.CANCELLED,
        MobileHermesStreamTerminal.COMPLETED,
        MobileHermesStreamTerminal.STREAMING,
        -> true
        MobileHermesStreamTerminal.ERROR -> false
    }

    fun marksError(terminal: MobileHermesStreamTerminal): Boolean = terminal == MobileHermesStreamTerminal.ERROR

    fun shouldDropEmptyAssistant(text: String, toolCallCount: Int, isError: Boolean, terminal: MobileHermesStreamTerminal): Boolean = when (terminal) {
        MobileHermesStreamTerminal.STOPPED,
        MobileHermesStreamTerminal.CANCELLED,
        -> text.trim().isEmpty() && toolCallCount == 0 && !isError
        else -> false
    }

    fun shouldApplyChunk(chunkThreadId: String?, activeThreadId: String?, chunkGeneration: Int, activeGeneration: Int): Boolean {
        if (chunkGeneration != activeGeneration) return false
        val chunk = chunkThreadId?.trim().orEmpty()
        val active = activeThreadId?.trim().orEmpty()
        if (chunk.isEmpty() || active.isEmpty()) return false
        return chunk == active
    }

    fun shouldAppendUserMessage(lastRole: String?, lastText: String?, incomingText: String, reason: String): Boolean {
        val incoming = incomingText.trim()
        if (incoming.isEmpty()) return false
        if (reason == "reconnect" &&
            lastRole == "user" &&
            lastText?.trim() == incoming
        ) {
            return false
        }
        return true
    }

    fun shouldRenderToolCalls(toolCallCount: Int, terminal: MobileHermesStreamTerminal): Boolean =
        toolCallCount > 0 && terminal != MobileHermesStreamTerminal.ERROR

    fun attachmentDisposition(id: String, mimeType: String, byteSize: Int, path: String): MobileHermesAttachmentDisposition {
        if (id.trim().isEmpty() || mimeType.trim().isEmpty() || path.trim().isEmpty() || byteSize < 0) {
            return MobileHermesAttachmentDisposition.REJECTED
        }
        return MobileHermesAttachmentDisposition.ACCEPTED
    }

    fun conversationDeepLink(threadId: String, exists: Boolean): MobileHermesConversationDeepLink {
        if (threadId.trim().isEmpty()) return MobileHermesConversationDeepLink.INVALID
        return if (exists) MobileHermesConversationDeepLink.LOADED else MobileHermesConversationDeepLink.MISSING
    }

    fun missingConversationMessage(outcome: MobileHermesConversationDeepLink): String? = when (outcome) {
        MobileHermesConversationDeepLink.LOADED -> null
        MobileHermesConversationDeepLink.MISSING -> "This conversation is no longer on this device."
        MobileHermesConversationDeepLink.INVALID -> "This conversation link is invalid."
    }
}

File: OpenBurnBarCore/Sources/OpenBurnBarKernel/SharedModels/MobileHermesConversationPolicy.swift
import Foundation

/// User-visible stream terminal. Cancel/stop are not errors.
public enum MobileHermesStreamTerminal: String, Sendable, Equatable {
    case streaming
    case stopped
    case cancelled
    case error
    case completed
}

public enum MobileHermesAttachmentDisposition: String, Sendable, Equatable {
    case accepted
    case rejected
}

public enum MobileHermesConversationDeepLink: String, Sendable, Equatable {
    case loaded
    case missing
    case invalid
}

/// Hermes/Pi stream, attachment, and assistant-link decisions.
/// Source: iOS `PiService.cancelStreaming` / `HermesService.cancelGeneration`.
public enum MobileHermesConversationPolicy {
    public static func terminal(forEvent event: String) -> MobileHermesStreamTerminal {
        switch event {
        case "stop":
            return .stopped
        case "cancel":
            return .cancelled
        case "error":
            return .error
        case "complete":
            return .completed
        default:
            return .error
        }
    }

    public static func keepsPartial(_ terminal: MobileHermesStreamTerminal) -> Bool {
        switch terminal {
        case .stopped, .cancelled, .completed, .streaming:
            return true
        case .error:
            return false
        }
    }

    public static func marksError(_ terminal: MobileHermesStreamTerminal) -> Bool {
        terminal == .error
    }

    /// Empty assistant placeholders are dropped on stop/cancel so the chat
    /// does not keep a blank bubble. Tool-call turns are kept.
    public static func shouldDropEmptyAssistant(
        text: String,
        toolCallCount: Int,
        isError: Bool,
        terminal: MobileHermesStreamTerminal
    ) -> Bool {
        switch terminal {
        case .stopped, .cancelled:
            return text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                && toolCallCount == 0
                && !isError
        case .error, .completed, .streaming:
            return false
        }
    }

    /// A late chunk for thread A must not apply to thread B.
    public static func shouldApplyChunk(
        chunkThreadId: String?,
        activeThreadId: String?,
        chunkGeneration: Int,
        activeGeneration: Int
    ) -> Bool {
        guard chunkGeneration == activeGeneration else { return false }
        let chunk = chunkThreadId?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let active = activeThreadId?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if chunk.isEmpty || active.isEmpty { return false }
        return chunk == active
    }

    /// Reconnect of an in-flight turn must not duplicate the user row.
    public static func shouldAppendUserMessage(
        lastRole: String?,
        lastText: String?,
        incomingText: String,
        reason: String,
        hasAttachments: Bool = false
    ) -> Bool {
        let incoming = incomingText.trimmingCharacters(in: .whitespacesAndNewlines)
        if incoming.isEmpty && !hasAttachments { return false }
        if reason == "reconnect",
           lastRole == "user",
           (lastText ?? "").trimmingCharacters(in: .whitespacesAndNewlines) == incoming {
            return false
        }
        return true
    }

    public static func shouldRenderToolCalls(
        toolCallCount: Int,
        terminal: MobileHermesStreamTerminal
    ) -> Bool {
        toolCallCount > 0 && terminal != .error
    }

    public static func attachmentDisposition(
        id: String,
        mimeType: String,
        byteSize: Int,
        path: String
    ) -> MobileHermesAttachmentDisposition {
        let cleanId = id.trimmingCharacters(in: .whitespacesAndNewlines)
        let mime = mimeType.trimmingCharacters(in: .whitespacesAndNewlines)
        let location = path.trimmingCharacters(in: .whitespacesAndNewlines)
        if cleanId.isEmpty || mime.isEmpty || location.isEmpty || byteSize < 0 {
            return .rejected
        }
        return .accepted
    }

    public static func conversationDeepLink(
        threadId: String,
        exists: Bool
    ) -> MobileHermesConversationDeepLink {
        if threadId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return .invalid
        }
        return exists ? .loaded : .missing
    }

    public static func missingConversationMessage(
        _ outcome: MobileHermesConversationDeepLink
    ) -> String? {
        switch outcome {
        case .loaded:
            return nil
        case .missing:
            return "This conversation is no longer on this device."
        case .invalid:
            return "This conversation link is invalid."
        }
    }
}
